### PR TITLE
[BEAM-14153] Reshuffled Row Coder PCollection used as Side Input cause encoder issues.

### PR DESCRIPTION
**Pull request by lostluck**
Adding the iterable side inputs revealed a latent issue to the SDK's approach to Reshuffle.
Revealed by having a reshuffle pipe directly to a side input (such as used in a PAssert).
The built in reshuffle was leaning on PCollections for the element pcollections which is brittle on some runners. 
The solution was to serialize the full coders as a payload for the reshuffle transforms to avoid not having the information available and to guarantee we use the same coder for element encoding and decoding of the reshuffled elements.

Note, technically only using the same coder ID did work for the affected runners, however that relies on the original pipeline coder components to be available on all bundle descriptors, which is also brittle.

Adds regression tests for revealing the issue, and unit tests to add some (but not complete) coverage to the new reshuffle behavior.

Affected runners were Dataflow, Python Portable and the google internal runner. Flink and Spark were unaffected.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by lostluck** [comment]
Run Go Flink ValidatesRunner

**Comment by lostluck** [comment]
Run Go PostCommit

**Comment by lostluck** [comment]
Run Go Spark ValidatesRunner

**Comment by lostluck** [comment]
Run Go Samza ValidatesRunner

**Comment by lostluck** [comment]
Run Go Samza ValidatesRunner

**Comment by lostluck** [comment]
Run Go Flink ValidatesRunner

**Comment by lostluck** [comment]
Run Go PostCommit

**Comment by lostluck** [comment]
Using the preserved coder ID works, but is dis-satisfying since I don't think "each bundle includes all the coders" is a guarantee that we should lean on.
I need to adjust the payload to have a map of IDs to encoded coder, payloads and be able to reconstruct/extract the coder directly (likely with the existing CoderMarhsaller/CoderUnmarshaller helpers), which will avoid the issue entirely.

**Comment by lostluck** [comment]
Run Go Flink ValidatesRunner

**Comment by lostluck** [comment]
Run Go PostCommit

**Comment by lostluck** [comment]
R: @jrmccluskey 
This should unblock the imports.

File: sdks/go/pkg/beam/core/runtime/graphx/translate.go [hunk 1]
```diff
@@ -66,8 +65,7 @@ const (
 	URNLegacyProgressReporting = "beam:protocol:progress_reporting:v0"
 	URNMultiCore               = "beam:protocol:multi_core_bundle_processing:v1"
 
-	URNRequiresSplittableDoFn     = "beam:requirement:pardo:splittable_dofn:v1"
-	URNRequiresBundleFinalization = "beam:requirement:pardo:finalization:v1"
+	URNRequiresSplittableDoFn = "beam:requirement:pardo:splittable_dofn:v1"
```
Comment (damccorm) [inline sdks/go/pkg/beam/core/runtime/graphx/translate.go:68]: This looks like a bad merge with my bundle finalization changes

**Review by damccorm — COMMENTED** [review]
I did a high level overview and didn't see anything obviously wrong other than the bundle finalization bad merge, I'll leave the more detailed review for @jrmccluskey since he has more context here

**Review by jrmccluskey — APPROVED** [review]
LGTM mod Danny's catch of the bundle finalization URN getting clobbered.

File: sdks/go/pkg/beam/core/runtime/graphx/translate.go (hunk 1 above)
Comment (lostluck) [reply sdks/go/pkg/beam/core/runtime/graphx/translate.go]: Ah whoops!
I spent all my effort making sure the proto didn't get clobbered, I forgot about the rest of the change. Thanks!

**Pull request merged**